mchf-eclipse/drivers/audio/audio_convolution.h

```diff
@@ -0,0 +1,72 @@
+/*  -*-  mode: c; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4; coding: utf-8  -*-  */
+/************************************************************************************
+ **                                                                                **
+ **                                        UHSDR                                   **
+ **               a powerful firmware for STM32 based SDR transceivers             **
+ **                                                                                **
+ **--------------------------------------------------------------------------------**
+ **                                                                                **
+ **  Description:   Code for fast convolution filtering                            **
+ **  Licence:		GNU GPLv3                                                      **
+ ************************************************************************************/
+
+/* Define to prevent recursive inclusion -------------------------------------*/
+#ifndef __AUDIO_CONVOLUTION_H
+#define __AUDIO_CONVOLUTION_H
+
+#include "arm_math.h"
+//#include "softdds.h"
+#include "uhsdr_hw_i2s.h"
+#include "uhsdr_board.h"
+#include "codec.h"
+#include "audio_driver.h"
+#include "radio_management.h"
+#include "usbd_audio_if.h"
+
+extern arm_fir_decimate_instance_f32   DECIMATE_RX_I;
+extern arm_fir_decimate_instance_f32   DECIMATE_RX_Q;
+//extern AudioDriverBuffer adb;
+
+#ifdef USE_CONVOLUTION
+#define FFT_CONVOLUTION_SIZE 256
+#define CONVOLUTION_MAX_NO_OF_BLOCKS 8
+#define CONVOLUTION_MAX_NO_OF_COEFFS 2048
+#endif
+
+#ifdef USE_CONVOLUTION
+
+typedef struct
+{
+    // for convolution filtering
+    float32_t				i_buffer_convolution[FFT_CONVOLUTION_SIZE / 2];
+    float32_t				q_buffer_convolution[FFT_CONVOLUTION_SIZE / 2];
+
+    float32_t				fmask[CONVOLUTION_MAX_NO_OF_BLOCKS][FFT_CONVOLUTION_SIZE * 2];
+    float32_t				fftin[FFT_CONVOLUTION_SIZE * 2];
+    float32_t				fftout[CONVOLUTION_MAX_NO_OF_BLOCKS][FFT_CONVOLUTION_SIZE * 2];
+    float32_t				ifftout[FFT_CONVOLUTION_SIZE * 2];
+    float32_t				accum[FFT_CONVOLUTION_SIZE * 2];
+    float32_t               a_buffer[2][FFT_CONVOLUTION_SIZE / 2]; // for convolution, we need an output buffer of 128 samples
+} ConvolutionBuffers;
+
+//extern ConvolutionBuffers cob;
+static ConvolutionBuffers cob;
+
```

Comment:
Never declare a static variable in a header unless you know what you are doing! This creates in every compiled .c file which includes audio_convolution.h a "cob" variable (but a different one for each file, since it is "static" not so not globally visible). No one said C is a fully logically structured language.